File: api/path_write.go
package vaku

import (
	"errors"
)

var (
	// ErrPathWrite when PathWrite errors.
	ErrPathWrite = errors.New("path write")
	// ErrVaultWrite when the underlying Vault API write fails.
	ErrVaultWrite = errors.New("vault write")
)

// PathWrite writes data to a path.
func (c *Client) PathWrite(p string, d map[string]interface{}) error {
	if d == nil {
		return newWrapErr(p, ErrPathWrite, ErrNilData)
	}

	vaultPath, mv, err := c.rewritePath(p, vaultWrite)
	if err != nil {
		return newWrapErr(p, ErrPathWrite, err)
	}

	if mv == mv2 {
		d = map[string]interface{}{
			"data": d,
		}
	}

	_, err = c.vl.Write(vaultPath, d)
	if err != nil {
		return newWrapErr(p, ErrPathWrite, newWrapErr(err.Error(), ErrVaultWrite, nil))
	}

	return nil
}

File: src/api/path_write.rs
//! `PathWrite`.

use crate::api::client::Client;
use crate::api::error::{Error, ErrorKind};
use crate::api::mounts::{rewrite_path, MountVersion, VaultOp};
use crate::api::secret::KV2_DATA;
use serde_json::{Map, Value};

impl Client {
    /// Write `data` to `p`. Mirrors Go's `PathWrite` — a `None` data argument
    /// yields [`ErrorKind::NilData`].
    pub async fn path_write(
        &self,
        p: &str,
        data: Option<Map<String, Value>>,
    ) -> Result<(), Error> {
        let Some(data) = data else {
            return Err(Error::wrap(
                p,
                ErrorKind::PathWrite,
                Some(Box::new(Error::wrap("", ErrorKind::NilData, None))),
            ));
        };

        let (vault_path, mv) = rewrite_path(self.src().mount_provider.as_ref(), p, VaultOp::Write)
            .await
            .map_err(|e| Error::wrap(p, ErrorKind::PathWrite, Some(Box::new(e))))?;

        let body: Value = if matches!(mv, MountVersion::Mv2) {
            let mut wrapped = Map::new();
            wrapped.insert(KV2_DATA.to_string(), Value::Object(data));
            Value::Object(wrapped)
        } else {
            Value::Object(data)
        };

        self.src().logical.write(&vault_path, body).await.map_err(|e| {
            Error::wrap(
                p,
                ErrorKind::PathWrite,
                Some(Box::new(Error::wrap(&e.to_string(), ErrorKind::VaultWrite, None))),
            )
        })?;
        Ok(())
    }
}
